addon/model.js

```diff
@@ -159,11 +159,29 @@ function disallowAliasSet(object, key, value) {
 
 
 /**
- * Given an array, representing the path to a nested model, construct a changed keys
- * structure, which express the changes to the overall structure and can be handled
- * by _notifyProperties correctly.
+ * Construct a list of changed keys, which `_notifyProperties` can use
+ * to invalidate the properties correctly, given an array, representing
+ * the path to the model and the list of changed keys.
  *
- * In case path is an empty array (the top model), then the `changedKeys` are returned.
+ * For example, given the following path and changed keys:
+ *  * ```
+ * // if the following paths are invalidated
+ * // foo.bar.baz.prop1
+ * // foo.bar.baz.prop2
+ * // the input should be:
+ * let path = ['foo', 'bar', 'baz'];
+ * let changedKeys = ['prop1', 'prop2'];
+ *
+ * // the resulting list should be:
+ * let result = constructChangedKeys(path, changedKeys);
+ * // result is:
+ * // ['foo', ['bar', ['baz', 'prop1', 'prop2']]]
```

Comment:
is this correct? I would have expected:

```
['foo', ['bar', ['baz', ['prop1', 'prop2']]]]
```

Reply:
It is a possible alternative, but the current code slices the array to get the changed key and the list of nested keys.

Reply:
what would happen if `bar` had also changed `biz` in addition to `baz.prop1` and `baz.prop2` ? `['foo', ['bar', ['biz' 'baz', 'prop1', 'prop2']]]` wouldn't work.

Reply:
The list of changed properties will be:
`['foo', ['bar', ['baz', 'prop1', 'prop1'], 'biz']]` or `['foo', ['bar', 'biz', ['baz', 'prop1', 'prop1']]]`

For the record, `constructChangedKeys` cannot generate such structure as it expects a path and list of changed keys, although the changed keys, might be something constructed with another call to `constructChangedKeys`.

Reply:
I see, so today we would end up with separate lists of changed keys per-path, not a union of those lists

Reply:
I would like to say 'yes', but I'm not sure it will completely right. The specific function we are discussing here deals with changed keys in single path, therefore you can say it generate separate lists per path, although it is more correct to say it generate single list always as the path is only one.

The overall structure encodes the changes without duplications, e.g. if there are multiple changed keys in single path `foo.bar`, the prefix `foo.bar` will appear once in the overall structure. However, this is only handled in the `merge` function.